tests/test_cases/simulation/test_mpi_ensemble.cu

```diff
@@ -0,0 +1,289 @@
+    validateLogs();
+}
+TEST_F(TestMPIEnsemble, error_slow) {
+    model->newLayer().addHostFunction(throw_exception);
+    initEnsemble();
+    ensemble->Config().error_level = CUDAEnsemble::EnsembleConfig::Slow;
+    // Capture stderr and stdout
+    testing::internal::CaptureStdout();
+    testing::internal::CaptureStderr();
+    // Expect no warnings (stderr) but outputs on progress and timing
+    if (world_rank == 0) {
+        EXPECT_THROW(ensemble->simulate(*plans), flamegpu::exception::EnsembleError);
+        // @todo can't capture total number of successful/failed runs
+        // Get stderr and stdout
+        const std::string output = testing::internal::GetCapturedStdout();
+        const std::string errors = testing::internal::GetCapturedStderr();
+        EXPECT_TRUE(output.find("CUDAEnsemble progress") != std::string::npos);   // E.g. CUDAEnsemble progress: 1/2
+        EXPECT_TRUE(output.find("CUDAEnsemble completed") != std::string::npos);  // E.g. CUDAEnsemble completed 2 runs successfully!
+        EXPECT_TRUE(errors.find("Warning: Run 10 failed on rank ") != std::string::npos);  // E.g. Warning: Run 10 failed on rank 0, device 0, thread 0 with exception:
+    } else {
+        // Only rank 0 raises exception
+        EXPECT_NO_THROW(ensemble->simulate(*plans));
+        // @todo can't capture total number of successful/failed runs
+        // Get stderr and stdout
+        const std::string output = testing::internal::GetCapturedStdout();
+        const std::string errors = testing::internal::GetCapturedStderr();
+        EXPECT_TRUE(output.empty());
+        EXPECT_TRUE(errors.empty());
+    }
+    // Existing logs should still validate
+    validateLogs();
+}
+TEST_F(TestMPIEnsemble, error_fast) {
+    model->newLayer().addHostFunction(throw_exception);
+    initEnsemble();
+    ensemble->Config().error_level = CUDAEnsemble::EnsembleConfig::Fast;
+    // Capture stderr and stdout
+    testing::internal::CaptureStdout();
+    testing::internal::CaptureStderr();
+    // Expect no warnings (stderr) but outputs on progress and timing
+    if (world_rank == 0) {
+        EXPECT_THROW(ensemble->simulate(*plans), flamegpu::exception::EnsembleError);
+        // @todo can't capture total number of successful/failed runs
+        // Get stderr and stdout
+        const std::string output = testing::internal::GetCapturedStdout();
+        const std::string errors = testing::internal::GetCapturedStderr();
+        EXPECT_TRUE(output.find("CUDAEnsemble progress") != std::string::npos);   // E.g. CUDAEnsemble progress: 1/2
+#ifdef _DEBUG
+        EXPECT_TRUE(errors.find("Run 10 failed on rank") != std::string::npos);   // E.g. Run 10 failed on rank 0, device 0, thread 0 with exception:
+#else
+        EXPECT_TRUE(errors.empty());
+#endif
+    } else {
+        // Only rank 0 raises exception
+        EXPECT_NO_THROW(ensemble->simulate(*plans));
+        // @todo can't capture total number of successful/failed runs
```

Comment:
Does this need addressing prior to merge?

Reply:
Hmm, I think this should be addressable. Will have a play.

Reply:
Oh this is by design, only rank 0 throws an exception/reports the number of failures etc. No point having all MPI runners pollute a log file with overlapping prints.